api/src/wfl/service/rawls.clj

```diff
@@ -37,14 +39,60 @@
    (create-snapshot-reference workspace snapshot-id name "")))
 
 (defn get-snapshot-reference
-  "Return the snapshot reference in fully-qualified Terra WORKSPACE with REFERENCE-ID."
+  "Return the snapshot reference in `workspace` with `reference-id`."
   [workspace reference-id]
-  (get-workspace-json workspace "snapshots" reference-id))
+  (get-workspace-json workspace snapshot-endpoint reference-id))
+
+(defn get-snapshot-references
+  "Lazily returns the snapshot references in `workspace`
+  with `limit` resources per page (default: 100)."
+  ([workspace limit]
+   (letfn
+    [(page [offset]
+       (let [{:keys [resources] :as _response}
+             (-> (workspace-api-url workspace snapshot-endpoint)
+                 (http/get {:headers      (auth/get-auth-header)
+                            :query-params {:offset offset
+                                           :limit  limit}})
+                 (util/response-body-json))
+             total    (+ offset (count resources))]
+         (lazy-cat resources (when (seq resources) (page total)))))]
+     (util/lazy-unchunk (page 0))))
+  ([workspace]
+   (get-snapshot-references workspace 100)))
+
+(defn create-or-get-snapshot-reference
+  "Return first snapshot reference for `snapshot-id` in `workspace`,
+  creating it as `name` with `description` if it does not yet exist."
+  ([workspace snapshot-id name description]
+   (try
+     (create-snapshot-reference workspace snapshot-id name description)
+     (catch ExceptionInfo cause
+       (let [{:keys [status] :as _data} (.getData cause)]
+         (cond
+           (= 409 status)
+           ;; On a duplicate resource exception, try to fetch the resource.
+           (if-let [reference
+                    (->> (get-snapshot-references workspace)
+                         (filter #(= (get-in % [:reference :snapshot])
+                                     snapshot-id))
+                         first)]
+             reference
```

Comment:
It would be nice to move this into a helper function to break the line count a bit. Then your catch expression could read something like
```clojure
(def ^:private reference-creation-failed-message
  (str/join " " ["Could not create snapshot reference and"
                 "found no snapshot reference matching the"
                 "snapshot id."]))

...

(when-not (== 409 status)
  (throw cause))
(or (get-reference-if-exists workspace snapshot-id)
    (throw (ex-info reference-creation-failed-message
                    {:workspace   workspace
                     :snapshot-id snapshot-id
                     :name        name}
             cause)))
```

Reply:
Good suggestion, will implement.